### ISSUE TITLE
Apply MaxConnectionsPerServer to HTTP/2?

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#40514]
SocketsHttpHandler.MaxConnectionsPerServer currently doesn't apply to HTTP/2.  In general with HTTP/2, the system behaves as if MaxConnectionsPerServer is 1 (since every request tends to reuse the same connection by spec and by design), and thus we implicitly respect MaxConnectionsPerServer (which will always be >= 1)... most of the time.  There are times, however, where the client will create a new connection to the server, such as if SocketsHttpHandler.PooledConnectionLifetime is set such that by the time the next request arrives, the connection is considered not usable for additional requests.  In such cases, we may actually end up with multiple live connections to the server, all of which are handling active request/responses, and thus could be violating MaxConnectionsPerServer if it was set to a low enough value.

Technically this is a bug. In practice, we could decide it's not worth enforcing.  But we should think through it, decide, and either close this or implement the gating.

**Comment by eiriktsarpalis** [comment]
I presume though that it will only become an issue (i.e. having > 2 live connections) for pathological values of `PooledConnectionLifetime`. Or am I missing something?

**Comment by stephentoub** [comment]
I don't think it requires a pathological value. Connections can be open for a very long time, and with bidirectional communication (e.g. gRPC), so can individual requests. It's also possible to hit I think if the server sends a goaway allowing existing requests to complete but requesting that no new ones are made on that collection.

**Comment by scalablecory** [comment]
How do we enforce this on HTTP1.1? Is it literally _connections_, or does it include e.g. pipelined requests?

**Comment by stephentoub** [comment]
> Is it literally connections

Yes

> or does it include e.g. pipelined requests?

We do not support HTTP/1.1 pipelining.

**Comment by scalablecory** [comment]
I think we do need this setting.

A server can send `GOAWAY`, be very slow to finish up existing streams, and force us to create a large number of connections for subsequent requests.

**Comment by eiriktsarpalis** [comment]
Given that SocketsHttpHandler instances can be used to manage both http and http2 connections at the same time, if we do decide to address this, I feel that we should probably expose a dedicated http2 property, i.e. `SocketsHttpHandler.MaxHttp2ConnectionsPerServer`. This would let us use a value that is sensible for http2 without impacting http1.1 behaviour.

**Comment by karelz** [comment]
Triage: We want to have a plan at least in 5.0. We need docs on that then.
